[SQL][SPARK-6785] fix DateUtils.fromJavaDate(java.util.Date) for Dates before 1970

ckadner opened the pull request:
BUG: With the current implementation, the from-to-Java date conversion will be off by one day for Dates before 1970 because of a rounding (truncate) flaw in the function DateUtils#millisToDays(Long).

FIX: The fix is to do the conversion using Double and flooring the fractions instead of truncating them.

NOTE: Before this fix, the code DID work for Dates that are NOT one millisecond before or after midnight in the system's local time zone.

ckadner committed:
SPARK-6785: use Math.floor() when converting milliseconds to days in function millisToDays(Long) otherwise the result of calling function fromJavaDate(java.util.Date) will be off by one day for Dates before 1970

AmplabJenkins commented:
Can one of the admins verify this patch?

davies commented:
OK to test

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala:
```diff
@@ -41,8 +41,9 @@ object DateUtils {
-  def millisToDays(millisLocal: Long): Int = {
-    ((millisLocal + LOCAL_TIMEZONE.get().getOffset(millisLocal)) / MILLIS_PER_DAY).toInt
+  def millisToDays(millisUtc: Long): Int = {
+    // SPARK-6785: use Math.floor so negative number of days (dates before 1970) work as input for function toJavaDate(Int)
```
This line is too long

davies commented:
Could you add `[SQL]` in the title? also follow the rules for title as others.

SparkQA commented:
[Test build #824 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/824/consoleFull) for   PR 6242 at commit [`6bb4ff4`](https://github.com/apache/spark/commit/6bb4ff4cbfc00b630504e23442f3137500743e8e).

SparkQA commented:
[Test build #824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/824/consoleFull) for   PR 6242 at commit [`6bb4ff4`](https://github.com/apache/spark/commit/6bb4ff4cbfc00b630504e23442f3137500743e8e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

davies commented:
@ckadner Thanks for working on this, the patch looks good to me, just one minor comment.

ckadner committed:
extract local variable to reduce line length

-    Math.floor((millisUtc.toDouble + LOCAL_TIMEZONE.get().getOffset(millisUtc)) / MILLIS_PER_DAY).toInt

+    val millisLocal = millisUtc.toDouble + LOCAL_TIMEZONE.get().getOffset(millisUtc)
+    Math.floor(millisLocal / MILLIS_PER_DAY).toInt

ckadner commented:
Thanks for your review Davies! I corrected the title and split the offending long line of code in 2.

davies commented:
LGTM, waiting for tests.

SparkQA commented:
[Test build #826 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/826/consoleFull) for   PR 6242 at commit [`7f3f9fb`](https://github.com/apache/spark/commit/7f3f9fbe1975bfbae150764adc115819eb03534d).

SparkQA commented:
[Test build #826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/826/consoleFull) for   PR 6242 at commit [`7f3f9fb`](https://github.com/apache/spark/commit/7f3f9fbe1975bfbae150764adc115819eb03534d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

ckadner committed:
public method must have explicit type

/home/jenkins/workspace/NewSparkPullRequestBuilder/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvaluationSuite.scala:420:8: Public method must have explicit type

ckadner commented:
Apologies for missing the explicit method return type. I just corrected that.

SparkQA commented:
[Test build #827 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/827/consoleFull) for   PR 6242 at commit [`ca999e1`](https://github.com/apache/spark/commit/ca999e1a5729b09949ef52e24b17744add4e9e6b).

SparkQA commented:
[Test build #827 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/827/consoleFull) for   PR 6242 at commit [`ca999e1`](https://github.com/apache/spark/commit/ca999e1a5729b09949ef52e24b17744add4e9e6b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

davies commented:
The unit test is outdated, please update it: ScalaReflectionRelationSuite: query case class RDD

ckadner committed:
[SPARK-6785][SQL] fix test "query case class RDD"

DateUtils now work for Dates before 1970 and "new Date(12345)" is not in 1970 if your System timezone is west of London

ckadner commented:
Thank you Davies for hanging in there with me. And apologies that I did not catch this one myself. I just fixed the unit test and the remaining unit tests in Catalyst, Hive, SQL are running through for me.

rxin commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33066 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33066/consoleFull) for   PR 6242 at commit [`a9ef7ee`](https://github.com/apache/spark/commit/a9ef7ee8ed23a8e9fe741e221551130296af7a7e).

SparkQA commented:
[Test build #33066 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33066/consoleFull) for   PR 6242 at commit [`a9ef7ee`](https://github.com/apache/spark/commit/a9ef7ee8ed23a8e9fe741e221551130296af7a7e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33066/
Test FAILed.

ckadner commented:
Hi Reynold (@rxin)
can you help me make sense of the last build log? **_...fails MiMa tests_**  Does this look like a configuration error? I do not readily see a connection to my code changes or how I could run this test suite on my own to root out possible problems.
Thank you in advance!

SparkQA commented:
[Test build #832 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/832/consoleFull) for   PR 6242 at commit [`a9ef7ee`](https://github.com/apache/spark/commit/a9ef7ee8ed23a8e9fe741e221551130296af7a7e).

srowen commented:
@ckadner A MiMa failure means it thinks something about the change modified a public API. It can be subtle sometimes. Here I can't pick out the actual MiMa error though. Try `./dev/mima` locally to see if it complains about anything.

davies commented:
@ckadner  There is one more test should be updated: Date comparison test 2

SparkQA commented:
[Test build #832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/832/consoleFull) for   PR 6242 at commit [`a9ef7ee`](https://github.com/apache/spark/commit/a9ef7ee8ed23a8e9fe741e221551130296af7a7e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

ckadner commented:
@srowen ...so far I have no luck with running ./dev/mima 

@davies ...I propose to remove the test case _"Date comparison test 2"_ since the expression `"CAST(CAST(0 AS timestamp) AS date) > CAST(0 AS timestamp)"` will produce different results based on the system timezone of the machine running it, 

i.e. it will evaluate to `true` in time zones east of GMT, i.e. Berlin, Germany ...

```
 "CAST(0 AS timestamp)" --> "1 Jan 1970 00:00:00 GMT"
 "CAST(CAST(0 AS timestamp) AS date)" --> "1 Jan 1970 01:00:00 GMT"
```
... but to `false` west of GMT, i.e. San Francisco ...
```
 "CAST(0 AS timestamp)" --> "1 Jan 1970 00:00:00 GMT"
 "CAST(CAST(0 AS timestamp) AS date)" --> "31 Dec 1969 08:00:00 GMT"
```
Note, this conversion is only accurate down to the _day_ of the `Date` object, since DateUtils uses full days to convert dates and the back-and-forth timezone adjustment

ckadner commented:
@srowen, @davies, thank you for your pointers!  I will work further on to this tomorrow.

ckadner commented:
Sorry, it took me some time to get back.

Turns out Hive suffers from the same bug ([HIVE-10178](https://issues.apache.org/jira/browse/HIVE-10178)) which causes the test cases based on `HiveComparisonTest` to fail. The Hive bug was fixed in _Hive 1.2.0._ and the code (and fix) in Hive's `DateWritable.java` looks remarkably similar to Spark's `DateUtils`.

@srowen, @davies, I need your guidance:
- how (and how often) do fixes in Hive get consumed by Spark? 
- should we move on with this fix in Spark and tweak the `HiveComparisonTest` to tolerate the discrepancy until Spark consumes Hive 1.2.0 (or another branch/version) that has the [HIVE-10178](https://issues.apache.org/jira/browse/HIVE-10178) fix?

Thanks in advance!

srowen commented:
Right now Spark is on Hive 0.12 / 0.13 although it will become more possible to use later Hive in the next version. So, that means our Hive comparison tests are going to get more fragile I think, if Hive behavior is changing. Is there any way to write this to expect several possible correct behaviors?

ckadner committed:
[SPARK-6785][SQL] tolerate outstanding fix for HIVE-10178 in hive comparison tests for dates before 1970

ckadner commented:
@srowen, thanks for your advice!

I modified the `HiveQuerySuite` to execute different test cases based on the version of Hive that is used internally by Spark SQL. It's not very elegant, but this way the Date cast tests will work now and after Spark upgrades to use Hive 1.2.0 (or later).

marmbrus commented:
We will only ever run the compatibility tests with the execution version of Hive.  So, I don't think it'll get any more brittle, but we will need to update the list/tests when we upgrade (WIP PR here #6748).

In any case where hive is returning the wrong answer, we usually just add the test case to the blacklist (in HiveCompatibilitySuite).  Since this is just our own test case tough, I'd move it to SQLQuerySuite and use `checkAnswer` instead.

marmbrus commented:
ping, do you think you could bring this up to date?

ckadner commented:
Hi Michael, 
for the failing hive comparison tests `"Date cast"` & `"Date comparison test 2"` I will add equivalent test cases to `SQLQuerySuite` probably in a new PR. I take it you want me to remove these 2 broken tests from `HiveQuerySuite` (as opposed to commenting them out) and delete the corresponding golden answer files?
Do we ever bring test cases like these back? Should I open a new JIRA to reinstate these test cases once [SPARK-8064](https://issues.apache.org/jira/browse/SPARK-8064) is resolved and the execution version of Hive is 1.2 ?

marmbrus commented:
Yeah you can go ahead and remove then and the golden files.  I think your new tests will be sufficient and we wan't need to bring them back (HiveQueryTests are easier to write, but aren't inherently better than tests where you manually hard code the answers).

ckadner commented:
Thanks Michael, I will open a new PR with a cleaned up version of my fix.

ckadner commented:
I opened a new pull-request [PR-6983](https://github.com/apache/spark/pull/6983)
